File: src/mcp/src/lookup.ts
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import DatabaseConstructor, { Database as DatabaseHandle } from 'better-sqlite3';
import { Evaluation, GameState } from './types.js';
import { HashCalc } from './shogi.js';

export class LookupDatabase {
  private readonly statement;

  constructor(private readonly database: DatabaseHandle) {
    this.statement = this.database.prepare('SELECT value FROM results WHERE hash = ?');
  }

  evaluateHash(hash: string): Evaluation {
    const row = this.statement.get(hash) as { value: string } | undefined;
    return row ? parseDbValue(row.value) : { result: 'unknown' };
  }

  evaluateState(state: GameState): Evaluation {
    return this.evaluateHash(HashCalc.hashToString(HashCalc.calcHash(state)));
  }

  close(): void {
    this.database.close();
  }
}

export function openLookupDatabase(dbPath = getDefaultLookupDbPath()): LookupDatabase {
  return new LookupDatabase(new DatabaseConstructor(dbPath, { readonly: true, fileMustExist: true }));
}

export function getDefaultLookupDbPath(): string {
  const currentDir = path.dirname(fileURLToPath(import.meta.url));
  return path.resolve(currentDir, '../../solver/data/lookup.db');
}

export function parseDbValue(value: string): Evaluation {
  try {
    const parsed = JSON.parse(value) as unknown;
    return parseRawEvaluation(parsed);
  } catch {
    return { result: 'unknown' };
  }
}

export function parseRawEvaluation(raw: unknown): Evaluation {
  if (raw === 'Draw' || raw === 'Sennichite' || raw === '千日手') return { result: 'draw' };
  if (!raw || typeof raw !== 'object') return { result: 'unknown' };
  const data = raw as Record<string, unknown>;
  if (typeof data.Win === 'number') return { result: 'win', ply: data.Win };
  if (typeof data.Lose === 'number') return { result: 'lose', ply: data.Lose };
  if (data.Draw !== undefined || data.Sennichite !== undefined) return { result: 'draw' };
  return { result: 'unknown' };
}

File: src/mcp-go/lookup.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	_ "github.com/mattn/go-sqlite3"
)

type LookupDatabase struct {
	db *sql.DB
}

func OpenLookupDatabase(dbPath string) (*LookupDatabase, error) {
	db, err := sql.Open("sqlite3", dbPath+"?mode=ro")
	if err != nil {
		return nil, fmt.Errorf("データベースを開けません: %w", err)
	}
	return &LookupDatabase{db: db}, nil
}

func (ldb *LookupDatabase) EvaluateHash(hash string) Evaluation {
	row := ldb.db.QueryRow("SELECT value FROM results WHERE hash = ?", hash)
	var value string
	if err := row.Scan(&value); err != nil {
		return Evaluation{Result: EvalUnknown}
	}
	return parseDBValue(value)
}

func (ldb *LookupDatabase) EvaluateState(state GameState) Evaluation {
	hash := HashCalc.HashToString(HashCalc.CalcHash(state))
	return ldb.EvaluateHash(hash)
}

func (ldb *LookupDatabase) Close() {
	ldb.db.Close()
}

func GetDefaultLookupDBPath() string {
	dir, err := os.Getwd()
	if err != nil {
		return "../../solver/data/lookup.db"
	}
	return filepath.Join(dir, "../../solver/data/lookup.db")
}

func parseDBValue(value string) Evaluation {
	var raw interface{}
	if err := json.Unmarshal([]byte(value), &raw); err != nil {
		return Evaluation{Result: EvalUnknown}
	}
	return parseRawEvaluation(raw)
}

func parseRawEvaluation(raw interface{}) Evaluation {
	switch v := raw.(type) {
	case string:
		if v == "Draw" || v == "Sennichite" || v == "千日手" {
			return Evaluation{Result: EvalDraw}
		}
	case map[string]interface{}:
		if winPly, ok := v["Win"]; ok {
			if winPlyFloat, ok := winPly.(float64); ok {
				ply := int(winPlyFloat)
				return Evaluation{Result: EvalWin, Ply: &ply}
			}
		}
		if losePly, ok := v["Lose"]; ok {
			if losePlyFloat, ok := losePly.(float64); ok {
				ply := int(losePlyFloat)
				return Evaluation{Result: EvalLose, Ply: &ply}
			}
		}
		if _, ok := v["Draw"]; ok {
			return Evaluation{Result: EvalDraw}
		}
		if _, ok := v["Sennichite"]; ok {
			return Evaluation{Result: EvalDraw}
		}
	}
	return Evaluation{Result: EvalUnknown}
}